### ISSUE TITLE
nested field encryption support

### ISSUE THREAD

**Issue by caltuntas** [wheresvic/mongoose-field-encryption#34]
Hello,

Firstly I want to thank you for this nice library. At the moment, library supports only outer level field encryption like in the PostSchema example you added. It can only encrypt "references" field as whole. But when only one of the fields inside "references" should be encrypted, it doesn't work.

So It would be nice to support nested field encryption with "dot notation" used in mongoDB itself.

[Query on Nested Field ](https://docs.mongodb.com/manual/tutorial/query-embedded-documents/#specify-equality-match-on-a-nested-field)

 For example , I should be able to define nested field as below 

`PostSchema.plugin(mongooseFieldEncyption, { fields: ["message", "references.author"], secret: "some secret key" });`

and it should only encrypt "references.author" not whole "references" field

**Comment by caltuntas** [comment]
After having a look at the closed issues I realized there was a similar issue https://github.com/wheresvic/mongoose-field-encryption/issues/1 (and merge request at the end) closed due to inactivity. It would be useful to add this feature.

**Comment by wheresvic** [comment]
@caltuntas I'll have a look at this again. The merge request you reference actually incorrectly mentioned the issue in question.

I cannot guarantee how fast I'll be able to develop this but let's say after the holidays would be a fair bet :)

**Comment by caltuntas** [comment]
Fair enough :) If I can find some time before you I may add this and make a pull request. Thanks

**Comment by wheresvic** [comment]
@caltuntas I had a look at this issue again and while implementing it would be possible, it would vastly complicate the code (and potentially break the current encrypted field naming convention).

I personally think that if you require nested field encryption, you could consider having sub-documents and apply the plugin on them and let mongoose itself handle the magic.

What do you think?

**Comment by rickmacgillis** [comment]
@wheresvic's suggestion on using subdocuments works perfectly. For those of you looking for a simple code example, consider the following.

```
const mongoose = require('mongoose');
const mongooseFieldEncryption = require("mongoose-field-encryption").fieldEncryption;

const CredentialSchema = new mongoose.Schema({
    type: {
        required: true,
        type: String,
    },
    value: {
        required: true,
        type: String,
    },
});

CredentialSchema.plugin(mongooseFieldEncryption, {
    fields: ["value"],
    secret: process.env.MONGOOSE_ENCRYPTION_KEY,
});

const accountSchema = new mongoose.Schema({
    provider: {
        type: String,
        required: true,
        lowercase: true,
        trim: true,
    },
    credentials: [CredentialSchema],
    owner: {
        required: true,
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
    },
});

module.exports = mongoose.model('Account', accountSchema);
```
That's a stripped down version of what I'm using to store OAuth credentials for multiple service providers for multiple users on an open source project I'm currently working on.

Hopefully this code helps someone.

**Comment by mexusbg** [comment]
+ 1 for this request.
@rickmacgillis suggestion works, but I would be nice if implemented

**Comment by 774649283** [comment]
`const mongoose = require("mongoose");
const schema = new mongoose.Schema({
    fieldA: { // 字段A
        type: String
    },
    fieldBArrays: [new mongoose.Schema({ // 字段B嵌套文档
        child: { // 子字段
            type: String
        }
    })]
});
const mongooseFieldEncryption = require('mongoose-field-encryption').fieldEncryption;
schema.plugin(mongooseFieldEncryption, {
    fields: ["fieldA","fieldBArrays.$.child"],
    secret: "秘钥",
    saltGenerator: function (secret) {
        return "1234567890123456"; //理想情况下，应使用该机密返回长度为16的字符串
    }
});`

I did this

**Comment by Yzhibin** [comment]
I experienced some issue with a similar structure like "Account schema has an array of Credential Schema, in which a field is marked as encrypted". After I use `find()` `findOne()` to get account(s), update some other fields but not the `credentials`, I will get and error of
```
"Cannot create field '-1' in element in { credentials: [...] }"
```
when I call `account.save()`

It seems like only happens to Array type, and I found using `account.markModified('credentials') solves this problem. 

I'm not sure if this behaviour is mentioned anywhere but since it is kind of related to the implementation described in this issue, I though I'd just post it here.

**Comment by simonestaffa** [comment]
I tried following the suggestion proposed by @rickmacgillis but it's not working on my side.

I have an UserSchema where I need to encrypt the "name" and "surname" fields. Then this schema contains a field "extra" which is based on another schema (as it is a nested object).
As suggested above, I have created the UserExtraSchema to allow nested fields encryption.

Unfortunately, I'm having some issues with this scenario. Therefore, the fields specified in the UserSchema are successfully encrypted, while the ones in the UserExtraSchema are not.

Below a snippet showing what I'm trying to do.

```
const UserExtraSchema: Schema = new Schema({
        city: {type: String},
        country: {type: String},
        address: {type: String},
        postalCode: {type: String}
});        

UserExtraSchema.plugin(mongooseFieldEncryption.fieldEncryption, {
    fields: ['address'],
    secret: process.env.MY_SECRET,
    saltGenerator(secret): string {
        return bcrypt.hashSync(secret, 10).substring(0, 16);
    }
});

const UserSchema: Schema = new Schema({
        _id: {type: String, required: true},
        name: {type: String, required: true},
        surname: {type: String, required: true},
        email: {type: String, required: true},
        extra: UserExtraSchema,
    }, {collection: 'users'}
);

UserSchema.plugin(mongooseFieldEncryption.fieldEncryption, {
    fields: ['name', 'surname],
    secret: process.env.MY_SECRET,
    saltGenerator(secret): string {
        return bcrypt.hashSync(secret, 10).substring(0, 16);
    }
});
```

**Comment by wheresvic** [comment]
Added test: https://github.com/wheresvic/mongoose-field-encryption/blob/036917d580e0d43c28f3331d0025af439d6ee18c/test/test-db.js#L212

**Comment by mexusbg** [comment]
> I tried following the suggestion proposed by @rickmacgillis but it's not working on my side.
> 
> I have an UserSchema where I need to encrypt the "name" and "surname" fields. Then this schema contains a field "extra" which is based on another schema (as it is a nested object). As suggested above, I have created the UserExtraSchema to allow nested fields encryption.
> 
> Unfortunately, I'm having some issues with this scenario. Therefore, the fields specified in the UserSchema are successfully encrypted, while the ones in the UserExtraSchema are not.
> 
> Below a snippet showing what I'm trying to do.
> 
> ```
> const UserExtraSchema: Schema = new Schema({
>         city: {type: String},
>         country: {type: String},
>         address: {type: String},
>         postalCode: {type: String}
> });        
> 
> UserExtraSchema.plugin(mongooseFieldEncryption.fieldEncryption, {
>     fields: ['address'],
>     secret: process.env.MY_SECRET,
>     saltGenerator(secret): string {
>         return bcrypt.hashSync(secret, 10).substring(0, 16);
>     }
> });
> 
> const UserSchema: Schema = new Schema({
>         _id: {type: String, required: true},
>         name: {type: String, required: true},
>         surname: {type: String, required: true},
>         email: {type: String, required: true},
>         extra: UserExtraSchema,
>     }, {collection: 'users'}
> );
> 
> UserSchema.plugin(mongooseFieldEncryption.fieldEncryption, {
>     fields: ['name', 'surname],
>     secret: process.env.MY_SECRET,
>     saltGenerator(secret): string {
>         return bcrypt.hashSync(secret, 10).substring(0, 16);
>     }
> });
> ```

I have the same issue

**Comment by rcpassos** [comment]
Hi, I have the same issue that @simonestaffa and @mexusbg

**Comment by wheresvic** [comment]
@mexusbg there is a test that uses exactly this example and it is working fine: https://github.com/wheresvic/mongoose-field-encryption/blob/036917d580e0d43c28f3331d0025af439d6ee18c/test/test-db.js#L212

the only difference I see from the example code is the way the `saltGenerator` function is setup. Although I do not see anything wrong with the one that you are using. The other point is that I do not see how you save and retrieve the document. Are you saving the main user document?

**Comment by mexusbg** [comment]
@wheresvic 
I'm saving the main user document.
User.create({})
User.findOneAndUpdate(findQuery,updateQuery)

**Comment by ctranstrum** [comment]
I have the same problem with subschema encryption when I use findOneAndUpdate() on the parent record, but when I switched to save(), it works.

**Comment by dor-flare** [comment]
hey , can any one help me encrypt the ssn field :


@Schema()
class CustomerQueryDetails {
  @Prop()
  dob: string;

  @Prop()
  firstName: string;

  @Prop()
  lastName: string;

  @Prop()
  ssn?: string;

  @Prop()
  address: Address;
}


@Schema({ _id: true, id: true, timestamps: true })
export class CreditCheck {
  _id: Types.ObjectId;
  id: string;

  @Prop()
  customerQueryDetails: CustomerQueryDetails;

  @Prop({ required: false })
  resultsData: ResultsData;

  @Prop({ required: true })
  resultStatus: boolean;

  @Prop({ required: false })
  errorCode: ErrorCode;

  @Prop({ required: true, index: true })
  customerId: string;

  @Prop()
  createdAt: Date;
}

**Comment by jbreadner** [comment]
There's been nothing on this thread for quite a while, but maybe this will help someone that was having the same problem as me.

The test referenced above clearly shows that when you do something like this:

```
document.save().then(found => console.log(found))
```

the "found" document (the one returned by save()) is encrypted.  This is not the behavior I expected; I expected the document to be encrypted on disk, but the returned document to be decrypted.

To solve this, I added a method to UserClass:

```
  saveAndDecrypt() {
    this.save().then(found => UserModel.findById(found._id));
  }
```

Now, this will do what I expect, save an encrypted document to disk, but leave the contents of the returned object unencrypted:

```
document.saveAndDecrypt().then(found => console.log(found))
```
